tower-http/src/cors.rs

```diff
@@ -0,0 +1,798 @@
+//!     .body(Body::empty())
+//!     .unwrap();
+//!
+//! let response = service
+//!     .ready()
+//!     .await?
+//!     .call(request)
+//!     .await?;
+//!
+//! assert_eq!(
+//!     response.headers().get(header::ACCESS_CONTROL_ALLOW_ORIGIN).unwrap(),
+//!     "*",
+//! );
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+
+use futures_core::ready;
+use http::{
+    header::{self, HeaderName, HeaderValue},
+    request::Parts,
+    Method, Request, Response, StatusCode,
+};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`Cors`] which adds headers for [CORS][mdn].
+///
+/// See the [module docs](crate::cors) for an example.
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+#[derive(Debug, Clone)]
+pub struct CorsLayer {
+    allow_credentials: Option<HeaderValue>,
+    allow_headers: Option<HeaderValue>,
+    allow_methods: Option<HeaderValue>,
+    allow_origin: Option<AnyOr<Origin>>,
+    expose_headers: Option<HeaderValue>,
+    max_age: Option<HeaderValue>,
+}
+
+const WILDCARD: &str = "*";
+
+impl CorsLayer {
+    /// Create a new `CorsLayer`.
+    ///
+    /// This creates a restrictive configuration. Use the builder methods to
+    /// customize the behavior.
+    pub fn new() -> Self {
+        Self {
+            allow_credentials: None,
+            allow_headers: None,
+            allow_methods: None,
+            allow_origin: None,
+            expose_headers: None,
+            max_age: None,
+        }
+    }
+
+    /// A very permissive configuration suitable for development:
+    ///
+    /// - Credentials allowed.
+    /// - All request headers allowed.
+    /// - All methods allowed.
+    /// - All origins allowed.
+    /// - All headers exposed.
+    /// - Max age set to 1 hour.
+    ///
+    /// Note this is not recommended for production use
+    pub fn permissive() -> Self {
+        Self::new()
+            .allow_credentials(true)
+            .allow_headers(Any)
+            .allow_methods(Any)
+            .allow_origin(Any)
+            .expose_headers(Any)
+            .max_age(Duration::from_secs(60 * 60))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Credentials`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    ///
+    /// let layer = CorsLayer::new().allow_credentials(true);
+    /// ```
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+    pub fn allow_credentials(mut self, allow_credentials: bool) -> Self {
+        self.allow_credentials = Some(allow_credentials.to_string().parse().unwrap());
+        self
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Headers`][mdn] header.
```

Comment:
By the way, this header is required when client sends `Access-Control-Request-Headers` (normally browser)
It would be good to add note for that to encourage users to set it

Reply:
Note added in 9f09fc55bf74b